chore(next.clayui.com/composing): add composing documentation (#2585)
liferay/clay: next.clayui.com/content/docs/foundations/composing.md

</div>

## Philosophy

Assuming that we want our components to be reusable and enable developers to have the flexibility to build on their own use cases while also helping developers who want quickly implement [Lexicon](https://lexicondesign.io) components, we have to consider the following:

Teams are constantly advancing and new features are always emerging that require new interactions and behavioral changes, a product changes a lot over time, being refined, aligning interests and changing behaviors in the UI.

Designers cannot get caught up in strict rules this blocks the creative mind and can prevent UI advances and interaction enhancements. The same goes for components, it has to allow developers to move forward with the standards made by designers and allow them to reuse components in the future.

Dealing with this is complex and requires balancing component flexibility and specialization, getting to that balance is difficult and we can spend a lot of time polishing and refining components.

> We are delivering Clay as a **Multi-Layered API library**: **low-level**, **high-level** and **pure html**. Each approach has its purposes of use:

Reviewer: I think we should add this just for clarity

```suggestion
We are delivering Clay as a **Multi-Layered API library**: **low-level**, **high-level** and **pure html**(@clayui/css). Each approach has its purposes of use:
```

Author: Yeah, sounds good

description: "Component composition philosophy"
---

<div class="nav-toc-absolute">
<div class="nav-toc">

- [Composition](#composition)
- [Philosophy](#philosophy)
    - [Low-level components](#low-level-components)

> 	- [High-level components](#high-level-components)

Reviewer: Indentation is off here. I think if your editor is configured to used spaces over tabs, it gets wonky when prettier formats.

</div>
</div>

> Clay follows some fundamentals for creating new components and as we distribute the components so that they can be used, the components follow some of the main React patterns: [composition](https://reactjs.org/docs/composition-vs-inheritance.html) or you can listen around for "bulding blocks".

Reviewer: ```suggestion
Clay follows some fundamentals for creating new components by following a common React pattern of 
[composition](https://reactjs.org/docs/composition-vs-inheritance.html).
```

Reviewer: I don't think so. "listen around" doesn't make much sense for this context and "building blocks" is also similar to composition and low-level, so I think it'd be best to leave it out.

Author: Hmm, ok. Np.

> We believe that using composition is the most correct and is the most widely used standard in the community. We want to allow people to reuse our components for different purposes, we understand that we have different people with different opinions on how we want to use Clay for their purposes. We think about how we can have a little of both worlds.

Reviewer: ```suggestion
We believe that using composition is helpful and is a widely used standard in the community. By using composition, it allows Clay to be widely used across various use cases.
```

## Composition

> The composition is very strong with React, JSX syntactic sugar makes it very powerful and more viable when coupled with special `children` prop for passing arbitrary child elements. Props in React components allows us to pass objects and consequently React elements since they are objects.

Reviewer: ```suggestion
The composition is very common in React because the JSX syntactic sugar makes it very powerful and more viable when coupled with the `children` prop for passing arbitrary child elements. Props in React components allows us to pass objects and consequently React elements since they are objects.
```

```jsx
<ClayModal>
	<ClayModal.Footer
		first={
			<ClayButton.Group spaced>
				<ClayButton displayType="secondary">
					{'Secondary'}
				</ClayButton>
				<ClayButton displayType="secondary">
					{'Secondary'}
				</ClayButton>
			</ClayButton.Group>
		}
		middle={
			<ClayButton>{'Primary'}</ClayButton>
		}
	/>
</ClayModal>
```

> So we use some of the composable component approach in React.

Reviewer: Not really sure what this sentence it trying to explain. Is this a title for the content below?

Author: Yeah, just one sentence to get into composing methods. Perhaps it would be more explicit:

```suggestion
So we follow some of React's composition approaches:
```

<div class="clay-ullist">

> 

Reviewer: might be helpful to have a title here

Author: Hmm 🤔, do you have any suggestions?

> - [**Containment**](https://reactjs.org/docs/composition-vs-inheritance.html#containment) - They are treated as generic blocks that accept the passing of arbitrary `children` to them, either using `children` props or the special naming convention itself.

Reviewer: ```suggestion
- [**Containment**](https://reactjs.org/docs/composition-vs-inheritance.html#containment) - Components that are treated as generic blocks that accept the passing of arbitrary `children` to them, either using `children` props or the special naming convention itself.
```

> - [**Specialization**](https://reactjs.org/docs/composition-vs-inheritance.html#specialization) - "Special case" components that are specific for some use purposes and not treated as something more "generic".

Reviewer: ```suggestion
- [**Specialization**](https://reactjs.org/docs/composition-vs-inheritance.html#specialization) - Special case components that are for a specific use case and not treated as something more generic.
```

</div>

## Philosophy

> Assuming that we want our components to be reusable and enable developers to have the flexibility to build on their use cases while helping developers who want to do something fast when just following [Lexicon](https://lexicondesign.io) specifications, we follow a few Thought lines.

Reviewer: ```suggestion
Assuming that we want our components to be reusable and enable developers to have the flexibility to build on their own use cases while also helping developers who want quickly implement [Lexicon](https://lexicondesign.io) components, we have to consider the following:
```

Teams are constantly advancing and new features are always emerging that require new interactions and behavioral changes, a product changes a lot over time, being refined, aligning interests and changing behaviors in the UI.

Designers cannot get caught up in strict rules this blocks the creative mind and can prevent UI advances and interaction enhancements. The same goes for components, it has to allow developers to move forward with the standards made by designers and allow them to reuse components in the future.

> Dealing with this is complex and requires balancing component flexibility and specialization, getting to that balance is difficult and we can spend a lot of time polishing and refining components to achieve **"component supremacy"**.

Reviewer: What even is "component supremacy"? Might be better to just leave that portion off.

```suggestion
Dealing with this is complex and requires balancing component flexibility and specialization, getting to that balance is difficult and we can spend a lot of time polishing and refining components.
```

Author: It's just a term to symbolize how difficult it would be to get the perfect balance of the component with regard to flexibility and specialization... Does that make sense to you?

Reviewer: It kind of does, but I think it might be more confusing to have it in than if we just remove that portion. I think we still convey the same thing by leaving it out.

Author: Ok, I think this is enough.

We are delivering Clay as a **Multi-Layered API library**: **low-level**, **high-level** and **pure html**. Each approach has its purposes of use:

<div class="clay-ullist">

> - **low-level** - Flexibility and use of first level composition so you can customize and create high level components.

Reviewer: ```suggestion
- **low-level** - Basic building blocks to provide flexibility so that you can customize and create high level components.
```

> - **high-level** - Speed ​​to build new features are usually specialized components and tend to cover only a few use cases, limiting their flexibility.

Reviewer: ```suggestion
- **high-level** - Highly specific component that tend to cover only specific use cases, limiting their flexibility.
```

> 

Reviewer: what about pure html? Im guessing this should just point to clay-css? Or we may not even need to include it since  this doc is specific to composition

Author: Hmm, not sure, maybe Patrick could work on a similar but specific document for clay-css.

</div>

#### Low-level components

> They are a great way to create new behaviors or extend, utilizing compositing to allow you to compose with other Clay components and build new possibilities.

Reviewer: ```suggestion
They are a great way to create new behaviors by utilizing composition to allow you to compose with other Clay components and build new possibilities.
```

> In essence, low-level components are more flexible but can slightly increase the "work" you have to do to get specifications, it requires you to do more compositions but that doesn't mean it's bad or the components are incomplete. Usually they tend to be [controlled components](https://reactjs.org/docs/uncontrolled-components.html).

Reviewer: ```suggestion
In essence, low-level components are more flexible but can slightly increase the "work" you have to do to get specifications, it requires you to do more composition but that doesn't mean it's bad or the components are incomplete. Usually they tend to be [controlled components](https://reactjs.org/docs/forms.html#controlled-components).
```

Clay follows a namespace pattern for low-level layer components, components are always nested according to responsibilities, common example:

<div class="clay-ullist">

- ClayDropDown
- ClayDropDown.Action
- ClayDropDown.Item
- ClayDropDown.ItemList
- ClayDropDown.Search

</div>

```jsx
<ClayDropDown
	active={expand}
	onActiveChange={setExpand}
	trigger={<ClayButton>{"Low-level"}</ClayButton>}
>
	<ClayDropDown.ItemList>
		<ClayDropDown.Item href="#1">Linkable</ClayDropDown.Item>
		<ClayDropDown.Item>
			<ClayCheckbox
				checked
				label="Checkbox"
				onChange={() => {}}
			/>
		</ClayDropDown.Item>
	</ClayDropDown.ItemList>
</ClayDropDown>
```

#### High-level components

> They are friendlier in simple use cases and help develop quickly but are specialized with specific cases. For their essence they are built on top of low-level components adding rules that help to reach [Lexicon](https://lexicondesign.io) specifications.

Reviewer: ```suggestion
They are friendlier in simple use cases and help develop quickly but are specialized for specific cases. They are built on top of low-level components adding rules that help to reach specific [Lexicon](https://lexicondesign.io) specifications.
```

> The peculiarity is that often high-level components follow "inheritance" rather than composition but this is not a rule that separates the two layers, it may happen to follow composition.

Reviewer: Not sure this line is very helpful, especially since we invalidate our "often" with the last clause of "it may happen to follow composition".

I think its best to just remove this line.

> Unlike low level components that tend to be controlled in many cases high level can be treated as [uncontrolled](https://reactjs.org/docs/uncontrolled-components.html).

Reviewer: Not sure this is necessarily true. We also try to also use controlled components for high-level. We do some internal wiring in some of them, but we still use onChange and value props typically.

Author: Yes, it's not 100% true but usually a high level tends to go in that direction, usually. I think there are cases with exception... but I think I will remove this so as not to create confusion.
